### PR TITLE
Split shop building in independent single job for optimization

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | Split docker building in an independent single job for optimization, then the sources and dump are shared as artifacts for the following steps that handles test commands thus making the setup of the shop faster<br><br>This PR comes with a related one in the core https://github.com/PrestaShop/PrestaShop/pull/33690 Not required but it allows optimizing a bit more the test phase (no need to install composer and node in the docker during the second test phase since everything has already been installed and built)<br/><br/>The workflow itself has multiple optimizations:<br><ul><li>Only one build for all following campaigns</li><li>Sanity tests are handled separately because they already handle installation, that saves some build time and improves stability</li><li>Keycloak service is only built for API tests</li><li>Long campaigns are started in priority to optimize the total duration</li><li>You can now run a single dedicated campaign</li><li>Javascript dependencies are run in parallel to win a bit of time, a systematic check is performed at the end to make sure the built is over</li><li>Docker images are pulled and built early in background while dependencies are built/installed to save some time during the final build</li></ul><br><br>New features:<ul><li>Possibility to run a single campaign</li><li>Possibility run force fast fail (still false by default)</li></ul>
| Type?             | improvement
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | ~
| Sponsor company   | ~
| How to test?      | ~

### Remaining possible improvement
- the base branch from PR could be used to avoid selecting it manually

# Benchmark

## Using the optimized PR https://github.com/PrestaShop/PrestaShop/pull/33690
Only this benchmark was run without another build in parallel for the sake of showing the improvement

#### Without cache
https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6207662484
Total duration: [31m 5s](https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6207662484/usage)
Total run time: 7h 20m 47s

#### With cache already set
https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6208145339
Total duration: [31m 18s](https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6208145339/usage)
Total run time: 7h 1m 43s

> [!NOTE]  
> In both cases sanity tests are fixed

#### Without cache
https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6278375332

#### With cache already set
https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6276944907
Total duration: [33m 30s](https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6276944907/usage)
Total run time: 7h 18m 49s

#### No cache system
https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6208570595
Total duration: [54m 49s](https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6208570595/usage)
Total run time: 10h 30m 48s

> [!WARNING]  
> Sanity tests failing, which makes the total duration longer because failing tests are long to finish

# Other PRs

## Merge PR
https://github.com/PrestaShop/PrestaShop/pull/33952
https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6208800734

## PR to update doctrine that failed sanity tests
https://github.com/PrestaShop/PrestaShop/pull/33459
https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6226478282

## PR targeting 8.1.x

https://github.com/PrestaShop/PrestaShop/pull/33930

### Split jobs
https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6226487438

### No split jobs
https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6222099607

## PR that updates ps_linklist
https://github.com/PrestaShop/PrestaShop/pull/33853

### Split jobs
https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6208823720

### No split jobs
https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/6222085952

### CODE REVIEW

**Comment by boubkerbribri** [comment]
I love this, did you start a test somewhere ?

**Comment by jolelievre** [comment]
@boubkerbribri yep one of the tests https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/5891548178 but it's not working yet I'm still experiencing

**Review by mflasquin — APPROVED** [review]

File: .github/workflows/pr_test.yml [hunk 1]
```diff
@@ -60,9 +60,161 @@ on:
...
+        timeout-minutes: 15
+        env:
+          URL_FO: ${{ ((github.event.inputs.base_branch == '8.0.x') || (github.event.inputs.base_branch == '1.7.8.x')) && 'http://localhost:8001/' || 'https://localhost:8002/' }}
+          VERSION: ${{ (github.event.inputs.base_branch == '1.7.8.x') && github.event.inputs.php_version || env.VERSION }}
+          # Initial build, we make assets and install shop data
+          DISABLE_MAKE: 0
+          PS_INSTALL_AUTO: 1
+        run: |
+          echo Build docker via docker composer
+          USER_ID=$(id -u) GROUP_ID=$(id -g) docker-compose -f docker-compose.yml up -d --build
```
Comment (M0rgan01) [inline .github/workflows/pr_test.yml:190]: by default it takes the docker-compose.yml file
```suggestion
          USER_ID=$(id -u) GROUP_ID=$(id -g) docker-compose up -d --build
```

File: .github/workflows/pr_test.yml [hunk 2]
```diff
@@ -181,131 +333,74 @@ jobs:
...
-          URL_FO: ${{ ((github.event.inputs.base_branch == '8.0.x') || (github.event.inputs.base_branch == '1.7.8.x')) && 'http://localhost:8001/' || 'https://localhost:8002/' }}
           VERSION: ${{ (github.event.inputs.base_branch == '1.7.8.x') && github.event.inputs.php_version || env.VERSION }}
+          # No install we force the sources and load the SQL dump
+          PS_INSTALL_AUTO: 0
+          DISABLE_MAKE: 1
         run: |
-          USER_ID=$(id -u) GROUP_ID=$(id -g) \
-          docker-compose -f docker-compose.yml up -d --build
+          echo Build the remaining dockers
+          USER_ID=$(id -u) GROUP_ID=$(id -g) docker-compose -f docker-compose.yml up -d --build
```
Comment (M0rgan01) [inline .github/workflows/pr_test.yml:400]:
```suggestion
          USER_ID=$(id -u) GROUP_ID=$(id -g) docker-compose up -d --build
```

File: .github/workflows/pr_test.yml [hunk 3]
```diff
@@ -181,131 +333,74 @@ jobs:
...
-
-      - name: Adapt Back Office layout
-        if: inputs.backoffice_layout == 'symfony'
+      - name: Setup database
         working-directory: ${{ env.PS_DIR }}
-        shell: bash
         run: |
-          echo PS_FF_SYMFONY_LAYOUT=true >> .env
+          echo Starting mysql docker alone
+          USER_ID=$(id -u) GROUP_ID=$(id -g) docker-compose -f docker-compose.yml up -d --build mysql
```
Comment (M0rgan01) [inline .github/workflows/pr_test.yml:381]:
```suggestion
          USER_ID=$(id -u) GROUP_ID=$(id -g) docker-compose up -d --build mysql
```

File: .docker/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,49 @@
+version: '2.1'
+
+services:
+    mysql:
+        image: prestashop-mysql-image
```
Comment (M0rgan01) [inline .docker/docker-compose.yml:5]: why not use the official mysql image? Same for keycloak

File: .docker/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            KEYCLOAK_ADMIN: ${KEYCLOAK_ADMIN:-admin}
+            KEYCLOAK_ADMIN_PASSWORD: ${KEYCLOAK_ADMIN_PASSWORD:-admin}
+        command:
+            - start-dev
+        ports:
+            - "8003:8080"
+
+    prestashop-git:
+        hostname: ${PS_HOSTNAME:-localhost}
+        image: prestashop-git-image
```
Comment (M0rgan01) [inline .docker/docker-compose.yml:25]: Is a dockerfile missing? or use the core one but I don't see a build

**Comment by jolelievre** [comment]
Comparison:
Regular single job: https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/5938546267 started at 14h09

**Review by Progi1984 — APPROVED** [review]

**Review by matks — CHANGES_REQUESTED** [review]
@jolelievre This PR needs a rebase 😢

**Review by boubkerbribri — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Comment by matks** [comment]
Great job @jolelievre and thank you @Progi1984 and @boubkerbribri for the extensive review work

**Pull request merged**